Add upgrade rules support for map PlayerReferences

penev92 opened the pull request:
Split off from #7873, because that needs to be able to upgrade the PlayerReferences in maps.
This just adds support for that. The other PR will add the first use-case.

I used [this commit](https://github.com/pchote/OpenRA/commit/cb32f8aa80542c7f059bc62df6f4bdc043260eaf) as example and now the PlayerReferences are loaded as MiniYAML by the map and kept that way.

Reviews should start with `Map.cs`, because that is where all the action is happening. Most of the rest is noise.

**Edit:** The action has moved to `MapPlayers.cs` now, but it all still starts from `Map.cs`.

pchote commented on OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:
```diff
@@ -1269,5 +1269,13 @@ internal static void UpgradeCursors(int engineVersion, ref List<MiniYamlNode> no
+		{
+			foreach (var node in nodes)
+			{
+				UpgradeCursors(engineVersion, ref node.Value.Nodes, node, depth + 1);
```
This probably isn't what you want.

pchote commented on OpenRA.Game/Server/Server.cs:
```diff
@@ -304,7 +304,7 @@ void ValidateClient(Connection newConn, string data)
 
 				if (client.Slot != null)
-					SyncClientToPlayerReference(client, Map.Players[client.Slot]);
+					SyncClientToPlayerReference(client, Map.GetPlayers()[client.Slot]);
```
You really should cache this instead of regenerating throw-away player definitions each time somebody joins.

pchote commented on OpenRA.Game/Map/Map.cs:
```diff
@@ -217,12 +217,24 @@ static object LoadVideos(MiniYaml y)
-		public int PlayerCount { get { return Players.Count(p => p.Value.Playable); } }
+		public int PlayerCount { get { return GetPlayers().Count(p => p.Value.Playable); } }
+
+		public Dictionary<string, PlayerReference> GetPlayers()
```
It would be nice if you could decouple this even further from Map.  Maybe create a separate MapPlayers class that you can initialize with a map and has a .Save() method.

pchote commented on OpenRA.Game/Map/Map.cs:
```diff
@@ -766,8 +772,10 @@ public void MakeDefaultPlayers()
+			var players = GetPlayers();
+
+			if (!players.ContainsKey("Neutral"))
+				players.Add("Neutral", new PlayerReference
```
It would be very useful for the new editor if this MakeDefaultPlayers method was also moved out of Map.  If you follow my suggestion about creating a new class for manipulating PlayerReferences then that would be a natural place to move this.

penev92 commented:
Added a `MapPlayers` class that handles player references and moved `MakeDefaultPlayers()` inside it.

pchote commented on OpenRA.Game/Map/Map.cs:
```diff
@@ -217,12 +216,18 @@ static object LoadVideos(MiniYaml y)
 		[FieldLoader.Ignore] public Lazy<Dictionary<string, ActorReference>> Actors;
 
-		public int PlayerCount { get { return Players.Count(p => p.Value.Playable); } }
+		public int PlayerCount
```
This may be better as a `Lazy<int>` (or perhaps as a wrapper to a private lazy property to avoid exposing the lazyness), so that this work only needs to be done once.

pchote replied:
Actually, this isn't even used.  Just delete it.

pchote commented on OpenRA.Game/Map/MapPreview.cs:
```diff
@@ -116,7 +116,7 @@ public void UpdateFromMap(Map m, MapClassification classification)
 			Type = m.Type;
 			Author = m.Author;
-			PlayerCount = m.Players.Count(x => x.Value.Playable);
+			PlayerCount = m.PlayerCount;
```
I guess this nullifies my "not even used" comment above (it *was* unused).

pchote replied:
It might be cleaner to move the yaml querying here instead?  Everything in the UI code that interacts with a map goes via the MapPreview anyway (because these support remote maps as well as local), and it would be nice to cut down some of the random functions off Map.

pchote commented on OpenRA.Game/Widgets/WidgetUtils.cs:
```diff
@@ -243,7 +243,8 @@ public static string ChooseInitialMap(string initialUid)
 					// Maps with bots disabled confuse new players
-					if (m.Map.Players.Any(s => !s.Value.AllowBots))
+					var players = new MapPlayers(m.Map.PlayerDefinitions);
+					if (players.Players.Any(s => !s.Value.AllowBots))
```
You could save some performace by caching this "allows bots" (or maybe even "is ideal map") value on the MapPreview.

pchote commented on OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs:
```diff
@@ -75,6 +75,8 @@ public bool InterpretCommand(S server, Connection conn, Session.Client client, s
 			if (!ValidateCommand(server, conn, client, cmd))
 				return false;
 
+			var players = new MapPlayers(server.Map.PlayerDefinitions).Players;
```
You should cache the MapPlayers object on Server (and update it whenever there is a change map order) instead of reparsing it in all of the server traits.

pchote commented on OpenRA.Mods.Common/UtilityCommands/LegacyMapImporter.cs:
```diff
@@ -123,7 +124,11 @@ public static Map Import(string filename, string mod, Ruleset rules, Action<stri
 			map.RequiresMod = mod;
-			map.MakeDefaultPlayers();
+
+			var players = new MapPlayers();
```
Isn't this just going to wipe out the players you have just imported?

penev92 replied:
My understanding of this is that you don't import players, but spawn points (at least D2k does it this way and the code there is the same).
After that you use `MakeDefaultPlayers()` to add the PlayerReference definitions to the map in order to be able to use those spawn points.
So yes, this creates an empty collection of players, but there is nothing to be deleted at this point anyway.

pchote replied:
See below.

penev92 replied:
Oh, I see. Will fix.

pchote commented:
This is a good start.  I have left a few more suggestions to improve it a bit more.

I will adapt my future pr that does similar changes for actors and smudges to follow the conventions we're setting up here.

penev92 commented:
Updated.

penev92 commented:
Decided to keep `public PlayerReference this[string key]`, unless someone has a problem with it.
It does make for a bit better-looking code in some places.

Phrohdoh commented:
Are you indexing with the race, the name, the internal name? etc.

penev92 commented:
I am keeping the current behavior, which is to index by `PlayerReference.Race` (changed to `PlayerReference.ID` in #7873).

pchote commented on OpenRA.Mods.Common/UtilityCommands/LegacyMapImporter.cs:
```diff
@@ -180,8 +185,10 @@ public void ConvertIniMap(string iniFile)
 
+			mapPlayers = new MapPlayers(map.PlayerDefinitions);
 			foreach (var p in players)
 				LoadPlayer(file, p, legacyMapFormat == IniMapFormat.RedAlert);
```
This is where we import players.

penev92 replied:
Ah, right.
So I could probably get away with `mapPlayers = new MapPlayers();` above, correct?

pchote commented on OpenRA.Mods.Common/UtilityCommands/LegacyMapImporter.cs:
```diff
@@ -180,8 +185,10 @@ public void ConvertIniMap(string iniFile)
+			mapPlayers = new MapPlayers();
 			foreach (var p in players)
 				LoadPlayer(file, p, legacyMapFormat == IniMapFormat.RedAlert);
+			map.PlayerDefinitions = mapPlayers.ToMiniYaml().ToList();
```
If you swap the order of waypoint and player parsing you could add MakeDefaultPlayers here and then drop the other block  completely.

penev92 replied:
Like this?

pchote commented:
Could you please incorporate the changes from 3af61428cb3c667462dead4a842450912b902dbb, and then squash everything down?

penev92 commented:
Done.

penev92 committed:
Reorganize PlayerReference loading to enable upgrade rules for them.

The PlayerReference definitions are now stored
as raw MiniYamlNodes in the map.

pchote commented on OpenRA.Game/Map/MapPlayers.cs:
```diff
@@ -0,0 +1,73 @@
+{
+	public class MapPlayers
+	{
+		public Dictionary<string, PlayerReference> Players;
```
This should be `readonly`.

pchote commented on OpenRA.Game/Map/MapPlayers.cs:
```diff
@@ -0,0 +1,73 @@
+			var firstRace = rules.Actors["world"].Traits
+				.WithInterface<CountryInfo>().First(c => c.Selectable).Race;
+
+			if (!Players.ContainsKey("Neutral"))
```
`Players` doesn't appear to have been initialized yet, so this will probably NRE?

pchote commented on OpenRA.Game/Server/Server.cs:
```diff
@@ -53,6 +53,7 @@ public class Server
 		public ServerSettings Settings;
 		public ModData ModData;
 		public Map Map;
+		public MapPlayers MapPlayers;
```
This is never assigned?

pchote commented on OpenRA.Mods.Common/ServerTraits/ColorValidator.cs:
```diff
@@ -175,8 +175,8 @@ public static bool ValidatePlayerNewColor(S server, Color askedColor, int player
-			var mapPlayerColors = server.Map.Players.Values
-				.Select(p => p.ColorRamp.RGB);
+			var players = new MapPlayers(server.Map.PlayerDefinitions).Players;
+			var mapPlayerColors = players.Values.Select(p => p.ColorRamp.RGB);
```
This should be able to use `server.MapPlayers` instead of creating its own.

pchote commented on OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs:
```diff
@@ -897,7 +897,9 @@ static Session.Slot MakeSlotFromPlayerReference(PlayerReference pr)
 			server.Map = server.ModData.MapCache[server.LobbyInfo.GlobalSettings.Map].Map;
-			server.LobbyInfo.Slots = server.Map.Players
+
+			server.MapPlayers = new MapPlayers(server.Map.PlayerDefinitions);
```
Oh, I see.  Strange that my IDE didn't reveal this when I searched for references.

pchote commented on OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:
```diff
@@ -1269,5 +1269,13 @@ internal static void UpgradeCursors(int engineVersion, ref List<MiniYamlNode> no
+		internal static void UpgradePlayers(int engineVersion, ref List<MiniYamlNode> nodes, MiniYamlNode parent, int depth)
+		{
+			foreach (var node in nodes)
+			{
```
Style nit: Unnecessary braces.

pchote commented on OpenRA.Game/Server/Server.cs:
```diff
@@ -53,6 +53,7 @@ public class Server
 		public ServerSettings Settings;
 		public ModData ModData;
 		public Map Map;
```
It would be worth adding an extra newline before this, and a comment `// Managed by LobbyCommands`.

penev92 commented:
Updated.

pchote commented on OpenRA.Game/Map/MapPlayers.cs:
```diff
@@ -0,0 +1,74 @@
+				Players.Add(p.Name, p);
+			}
+
+			Players.Add("Creeps", new PlayerReference
```
You can add this to the initializer after Neutral.

penev92 replied:
Not really, because it needs to add all other player names as enemies to this one.

pchote replied:
The order doesn't matter.

pchote replied:
that being something like `Enemies = Exts.MakeArray(playerCount, i => "Multi{0}".F(i))`

pchote commented on OpenRA.Mods.Common/UtilityCommands/LegacyMapImporter.cs:
```diff
@@ -180,14 +180,16 @@ public void ConvertIniMap(string iniFile)
 					.Select(kv => Pair.New(Exts.ParseIntegerInvariant(kv.Key),
 						LocationFromMapOffset(Exts.ParseIntegerInvariant(kv.Value), mapSize)));
 
+			mapPlayers = new MapPlayers(map.Rules, map.GetSpawnPoints().Length);
```
This block needs to be run *after* you import the waypoints :)

pchote commented:
Tested the behaviour ingame, the linter, legacy map import (and found a bug).  Haven't tested the legacy editor or d2k importer.

:+1: after you fix the map importer, and once someone with the ability to run them verifies that the d2k importer and legacy editor work.

penev92 commented:
Fixed.
For what is worth, both the D2k map importer and the external map editor work for me.

pchote commented on OpenRA.Mods.Common/UtilityCommands/LegacyMapImporter.cs:
```diff
@@ -193,19 +190,28 @@ public void ConvertIniMap(string iniFile)
-					a.Add(new LocationInit((CPos)kv.Second));
-					a.Add(new OwnerInit("Neutral"));
+					var a = new ActorReference("mpspawn")
+						{
```
-1 level of indention for these two please.

abcdefg30 commented:
Crashes when trying to import a map:
```
Received args: --map-import scg05eb.ini
System.ArgumentException: An item with the same key has already been added
(Ein Element mit dem gleichen Schlüssel wurde bereits hinzugefügt).
   at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
   at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
   at OpenRA.Mods.Common.UtilityCommands.LegacyMapImporter.LoadPlayer(IniFile file, String section, Boolean isRA)
   at OpenRA.Mods.Common.UtilityCommands.LegacyMapImporter.ConvertIniMap(String iniFile)
   at OpenRA.Mods.Common.UtilityCommands.LegacyMapImporter..ctor(String filename, Ruleset rules, Action`1 errorHandler)
   at OpenRA.Mods.Common.UtilityCommands.LegacyMapImporter.Import(String filename, String mod, Ruleset rules, Action`1 errorHandler)
   at OpenRA.Mods.Common.UtilityCommands.ImportLegacyMapCommand.Run(ModData modData, String[] args)
   at OpenRA.Utility.Program.Main(String[] args) in \OpenRA\OpenRA.Utility\Program.cs:Zeile 54.

penev92 commented:
That stacktrace makes no sense  (`LegacyMapImporter.LoadPlayer()` has nothing to do with dictionaries) and I successfully imported [this map](https://seacloud.cc/d/5d1adbebce/files/?p=/openra/scu08ea.ini), provided by @obrakmann, so I have no idea what that crash is.

abcdefg30 commented:
It also happened as I tried to import `scg03eb.ini`.

penev92 commented:
Fixed. The problem was some mission maps defined players that conflict with the default player names.

abcdefg30 commented:
Looks good to me now. :+1: / :white_check_mark:

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/9668138aae3499560084b36bce7568e787451587...031395c4e067457099f8aa87a0e5b7cb63f9bf02)
